birdhouse/config/thredds/default.env

```diff
@@ -6,10 +6,28 @@ export THREDDS_DOCKER=pavics/thredds-docker
 export THREDDS_IMAGE="${THREDDS_DOCKER}:${THREDDS_VERSION}"
 export THREDDS_ORGANIZATION="Birdhouse"
 
+# All default values for data locations are set to ensure backward compatibility
+export THREDDS_DATA_LOCATION_ON_CONTAINER='/pavics-ncml'
+export THREDDS_RAW_DATA_LOCATION_ON_CONTAINER='/pavics-data'
+export THREDDS_DATA_LOCATION_ON_HOST='${DATA_PERSIST_ROOT}/ncml'  # this default is for backward compatibility 
+export THREDDS_RAW_DATA_LOCATION_ON_HOST='${DATA_PERSIST_ROOT}/datasets'  # this default is for backward compatibility
+export THREDDS_DATA_LOCATION_NAME='Datasets'  # this default is for backward compatibility 
+export THREDDS_RAW_DATA_LOCATION_NAME='Birdhouse'  # this default is for backward compatibility 
+export THREDDS_DATA_LOCATION_NAME_lower='$(echo "$THREDDS_DATA_LOCATION_NAME" | tr "[:upper:]" "[:lower:]")'
+export THREDDS_RAW_DATA_LOCATION_NAME_lower='$(echo "$THREDDS_RAW_DATA_LOCATION_NAME" | tr '[:upper:]' '[:lower:]')'
```

Comment:
Even if the variables are for storing a lowercase representation, the names themselves should be kept all uppercase. 

I would like to know also why is lowercase enforced? I don't see any technical reason for this to be needed and it adds a lot of unnecessary variables that mostly duplicates configuration and `VARS`/`DELAYED_EVAL` references.

The only non-uppercase variant I find is the `name` property in the config XML. I believe this could be any-case with the same value as whichever value is set for these variables.

Reply:
> Even if the variables are for storing a lowercase representation, the names themselves should be kept all uppercase.

Fair.

> I would like to know also why is lowercase enforced? I don't see any technical reason for this to be needed and it adds a lot of unnecessary variables that mostly duplicates configuration and `VARS`/`DELAYED_EVAL` references.

These lower case variables are needed for maintaining full backward compatibility.

Reply:
I just realized, `THREDDS_DATA_LOCATION_NAME_lower` is misleading. 
If overridden, nothing actually enforces it to be lowercase... 
It would be better to simply have `THREDDS_DATA_LOCATION_DIR` or similar. 
And the default would simply be `THREDDS_DATA_LOCATION_DIR=datasets`.
Idem for "`RAW_DATA`".

Reply:
> I just realized, THREDDS_DATA_LOCATION_NAME_lower is misleading.
> If overridden, nothing actually enforces it to be lowercase...

Typically we have handled cases like this by not including variables like `THREDDS_DATA_LOCATION_NAME_lower` in the env.local.example file which indicates to the user that they are not meant to be overridden and are for internal use only.

Reply:
It is not ideal to have a "lower case" variable name from many perspectives, I agree with that. The issue is that to maintain full backward compatibility with what already exists in `config/thredds/catalog.xml.template` (as shown below) one need both naming case.

> `<datasetScan name="Birdhouse" ID="birdhouse" path="birdhouse" location="${THREDDS_RAW_DATA_LOCATION_ON_CONTAINER}">`
> `<datasetScan name="Datasets" ID="datasets" path="datasets" location="${THREDDS_DATA_LOCATION_ON_CONTAINER}">`

Reply:
I think Misha's suggestion works here.

Reply:
If I decide that I want my instance to use "CRIM" instead of "birdhouse", I could set `THREDDS_DATA_LOCATION_NAME=CRIM`. The current `THREDDS_DATA_LOCATION_NAME_lower` would then convert that to `crim`, so both directories and URLs will use `crim`. I have no choice to override `THREDDS_DATA_LOCATION_NAME_lower=CRIM` as well to obtain `/twitcher/ows/proxy/thredds/fileserver/CRIM/...`. 

`THREDDS_DATA_LOCATION_NAME_lower=CRIM` feels odd, and make uses of an "undocumented" `env.local.example` variable. So instead, I say we document them explicitly such that `default.env` and `env.local.example` both have: 
```env
# name of the directory used in THREDDS Catalog interface
export THREDDS_DATA_LOCATION_NAME=Birdhouse
# name of the directory in file system paths and URLs
export THREDDS_DATA_LOCATION_DIR=birdhouse
```

Reply:
If you want to control the name of the URL then sure, that makes sense. 

But just to be clear, in your example `THREDDS_DATA_LOCATION_DIR` is only used to set the URL path, not the name of the directory on the file system. That is currently set by the `THREDDS_DATA_LOCATION_ON_CONTAINER` variable: 

https://github.com/bird-house/birdhouse-deploy/pull/359/files#diff-4bd22da1236dbeeb19113dd6f1cec1bb094d578e8e4ae796309bc9b5c17f6958R34

If we're going to do it this way I'd suggest the more descriptive name `THREDDS_DATA_URL_PATH` 

We can change `THREDDS_DATA_LOCATION_ON_CONTAINER` to `THREDDS_DATA_LOCATION_DIR` as well

Reply:
I also agree that providing this flexibility makes sense. Let's make the following changes:

`THREDDS_DATA_LOCATION_NAME_lower -> THREDDS_DATA_URL_PATH`
`THREDDS_RAW_DATA_LOCATION_NAME_lower ->  THREDDS_RAW_DATA_URL_PATH`

I'd like to keep `THREDDS_DATA_LOCATION_ON_CONTAINER` as is to make it a clear counterpart of the variable `THREDDS_RAW_DATA_LOCATION_ON_HOST`. There are too many directories and volume mounts involved in the birdhouse stack and I think a nomenclature like this helps make sense of them while reducing the likelihood of errors.

Reply:
It is good to have the explicit distinction of `..._URL_PATH` and `..._LOCATION_ON_HOST` to offer the flexibility of each part. It is important to allow control over the exposed URL and the host path. 

I don't think the `..._LOCATION_ON_CONTAINER` is really that much important, since the purpose of THREDDS's `catalog.xml` is to remap any local (in-docker) path to some URL, so where the data actually resides in the container is irrelevant as long as it align with the configured volume path.

Reply:
> `THREDDS_DATA_LOCATION_NAME_lower -> THREDDS_DATA_URL_PATH`
> `THREDDS_RAW_DATA_LOCATION_NAME_lower -> THREDDS_RAW_DATA_URL_PATH`

Agreed, nicer name.  And they don't have to be delayed eval anymore right?   Documentation should be clear so the user can decide what to do with them (need lower case or not).

> I'd like to keep `THREDDS_DATA_LOCATION_ON_CONTAINER` as is to make it a clear counterpart of the variable `THREDDS_RAW_DATA_LOCATION_ON_HOST`.

Agreed too.


> I don't think the `..._LOCATION_ON_CONTAINER` is really that much important, since the purpose of THREDDS's `catalog.xml` is to remap any local (in-docker) path to some URL, so where the data actually resides in the container is irrelevant as long as it align with the configured volume path.

It is important if you are the admin of the host and you `docker exec` into the `thredds` container or you manage the files on disk (ex: to sync new files).  In this case, you do want to know the path in the container vs path on the real disk.

Another user that want to know the real path is the ncml developer.  All path referenced in an `.ncml` are location on container, example https://pavics.ouranos.ca/twitcher/ows/proxy/thredds/fileServer/datasets/reanalyses/day_ERA5_NAM.ncml, which correspond to this file https://github.com/Ouranosinc/pavics-vdb/blob/master/1-Datasets/reanalyses/day_ERA5_NAM.ncml (`.ncml` files in this repo are synced to our Thredds automatically)

So for the same `.nc` or `.ncml` file, the end users will care for the public URL on Thredds but the admin and developer will care for the actual location on disk and in the container.

Reply:
> I don't think the `..._LOCATION_ON_CONTAINER` is really that much important, since the purpose of THREDDS's `catalog.xml` is to remap any local (in-docker) path to some URL, so where the data actually resides in the container is irrelevant as long as it align with the configured volume path.

Exactly what @tlvu said. Also, in the absence of defining it in a variable we are just hardcoding it in the deployment everywhere it is needed, and that is not good.

Reply:
If that can make the life easier for users/dev/maintainers, then that is fine. 👍